go/private/go_repository.bzl

```diff
@@ -25,67 +25,84 @@ def _go_repository_impl(ctx):
   else:
     fail("neither commit or tag is specified", "commit")
 
-  # TODO(yugui): support submodule?
-  # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
-  remote = ctx.attr.remote
-  vcs = ctx.attr.vcs
-  importpath = ctx.attr.importpath
-  result = ctx.execute([
-      fetch_repo,
-      '--dest', ctx.path(''),
-      '--remote', remote,
-      '--rev', rev,
-      '--vcs', vcs,
-      '--importpath', importpath])
-  if result.return_code:
-    fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
+  if ctx.attr.url:
+    # explicit source url
+    if ctx.attr.vcs:
+        fail("cannot specify both of vcs and url")
+    ctx.download_and_extract(
+        url = ctx.attr.url,
+        sha256 = ctx.attr.sha256,
+        stripPrefix = ctx.attr.strip_prefix,
+        type = ctx.attr.type,
+    )
+  else:
+    # Using fetch repo
+    if ctx.attr.vcs and not ctx.attr.remote:
+      fail("if vcs is specified, remote must also be")
+    # TODO(yugui): support submodule?
+    # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
+    result = ctx.execute([
+        ctx.path(ctx.attr._fetch_repo),
+        '--dest', ctx.path(''),
+        '--remote', ctx.attr.remote,
+        '--rev', rev,
+        '--vcs', ctx.attr.vcs,
+        '--importpath', ctx.attr.importpath,
+    ])
+    if result.return_code:
+      fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
 
+  if (ctx.attr.disable_build_file_generation or 
+      ctx.path('BUILD').exists or 
```

Comment:
For case 1, I think there is no way to cope with it correctly, as you will need special processing to strip out BUILD files that bazel must not see as well as generate new ones, which may involve a custom tool to do a conversion, I think when that situation occurs, you are going to need a whole new rule to cope with it, or something more complicated than just running gazelle. I am not opposed to adding an extra flag when we find we need it, but I would really want an actual case to work on before I do.
I guess I could change ctx.attr.disable_build_file_generation to a tri state (off, auto, on) and default to auto?